pytorch_lightning/utilities/trainer_cli.py

```diff
@@ -0,0 +1,223 @@
+        Adds arguments from a datamodule class to a nested key of the parser
+
+        Args:
+            datamodule_class: A LightningDataModule class.
+            nested_key: Name of the nested namespace to store arguments.
+        """
+        assert issubclass(datamodule_class, LightningDataModule)
+        self.add_class_arguments(datamodule_class, nested_key)
+
+
+class SaveConfigCallback(Callback):
+    """Saves a LightningCLI config to the log_dir when training starts"""
+
+    def __init__(self, parser, config):
+        self.config_dump = parser.dump(config, skip_none=False)
+
+    def on_train_start(self, trainer, pl_module):
+        config_path = os.path.join(trainer.logger.log_dir, 'config.yaml')
+        with open(config_path, 'w') as outstream:
+            outstream.write(self.config_dump)
+
+
+class LightningCLI:
+    def __init__(
+        self,
+        model_class: Type[LightningModule],
+        datamodule_class: Type[LightningDataModule] = None,
+        save_config_callback: Type[Callback] = SaveConfigCallback,
+        trainer_class: Type[Trainer] = Trainer,
+        description: str = 'pytorch-lightning trainer command line tool',
+        parse_env: bool = False,
+        **kwargs
+    ):
+        """
+        Implementation of a simple configurable Trainer command line tool
+
+        Receives as input pytorch-lightning classes, which are instantiated using a
+        parsed configuration file or command line args and then runs trainer.fit.
+
+        Example, first implement the trainer.py tool as::
+
+            from mymodels import MyModel
+            from pytorch_lightning.utilities.jsonargparse_utils import LightningCLI
+            LightningCLI(MyModel)
+
+        Then in a shell, run the tool with the desired configuration::
+
+            $ python trainer.py --print-config > config.yaml
+            $ nano config.yaml  # modify the config as desired
+            $ python trainer.py --cfg config.yaml
+
+        Args:
+            model_class: The LightningModule class to train on.
+            datamodule_class: An optional LightningDataModule class.
+            save_config_callback: A callback class to save the training config.
+            trainer_class: An optional extension of the Trainer class.
+            description: Description of the tool shown when running --help.
+            parse_env: Whether environment variables are also parsed.
+            **kwargs: Additional arguments to instantiate Trainer.
+        """
+        if 'callbacks' not in kwargs:
+            kwargs['callbacks'] = []
+
+        self.model_class = model_class
+        self.datamodule_class = datamodule_class
+        self.save_config_callback = save_config_callback
+        self.trainer_class = trainer_class
+        self.trainer_kwargs = kwargs
+
+        self.init_parser(description, parse_env)
+        self.add_arguments_to_parser(self.parser)
+        self.add_core_arguments_to_parser()
+        self.parse_arguments()
```

Comment:
True, a hook after add_core_arguments_to_parser could be useful. Though the change of defaults I would personally do different. I would use the `default_config_files` option and have the defaults in a file that is installed with the package. Maybe we should also add `parser_kwargs` that would be used for arbitrary options to instantiate the parser.

Reply:
I have added `default_config_files` to the class. Though I haven't added a hook yet.

Reply:
I have added `before_parse_arguments` method to the cli class.

Reply:
As a side note, in jsonargparse since you have nested namespaces, setting defaults is done with a dictionary, e.g. `parser.set_defaults({'trainer.weights_summary': 'full'})`.